File: scripts/find-missing-ids-bulk.mjs
#!/usr/bin/env node
/**
 * Bulk download ALL PsyArXiv preprint IDs+titles from OSF API (concurrent),
 * then match against the 59 papers missing osf_ids.
 */
import { readFileSync, writeFileSync, existsSync } from 'fs';

const API_BASE = 'https://api.osf.io/v2/preprints/';
const PAPERS_JSON = '/home/z/my-project/psyarxiv-hub/data/papers.json';
const CACHE_FILE = '/home/z/my-project/scripts/psyarxiv-all-titles.json';
const OUTPUT_FILE = '/home/z/my-project/scripts/missing-osf-results.json';
const CONCURRENCY = 5;  // parallel API requests
const PAGE_SIZE = 100;
const MAX_PAGES = 200;  // safety limit

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

async function fetchJson(url, retries = 3) {
  for (let attempt = 0; attempt < retries; attempt++) {
    try {
      const resp = await fetch(url, {
        headers: { 'Accept': 'application/vnd.api+json' }
      });
      if (resp.ok) {
        return await resp.json();
      }
      if (resp.status === 429 && attempt < retries - 1) {
        await sleep((attempt + 1) * 3000);
        continue;
      }
      throw new Error(`HTTP ${resp.status}`);
    } catch (e) {
      if (attempt < retries - 1) {
        await sleep(1000 * (attempt + 1));
        continue;
      }
      throw e;
    }
  }
}

async function downloadAllTitles() {
  if (existsSync(CACHE_FILE)) {
    console.log(`Loading cached titles from ${CACHE_FILE}`);
    const cached = JSON.parse(readFileSync(CACHE_FILE, 'utf-8'));
    console.log(`Loaded ${cached.length} cached preprints`);
    return cached;
  }

  console.log('Downloading all PsyArXiv preprint titles...');

  // First, get the total count
  const firstPage = await fetchJson(
    `${API_BASE}?filter%5Bprovider%5D=psyarxiv&page%5Bsize%5D=1&sort=-date_created`
  );
  const total = firstPage.meta?.total || 10000;
  const totalPages = Math.ceil(total / PAGE_SIZE);
  console.log(`Total PsyArXiv preprints: ${total} (~${totalPages} pages)`);

  // Fetch all pages concurrently with rate limiting
  const allPreprints = [];
  let pagesFetched = 0;
  let activeRequests = 0;
  let pageQueue = [];

  for (let p = 1; p <= Math.min(totalPages, MAX_PAGES); p++) {
    pageQueue.push(p);
  }

  async function processPage(pageNum) {
    const url = `${API_BASE}?filter%5Bprovider%5D=psyarxiv&page%5Bsize%5D=${PAGE_SIZE}&page%5Bnumber%5D=${pageNum}&sort=-date_created`;
    const data = await fetchJson(url);
    const items = data.data || [];

    for (const item of items) {
      allPreprints.push({
        id: item.id,
        title: item.attributes?.title || '',
        date_created: (item.attributes?.date_created || '').substring(0, 10),
      });
    }

    pagesFetched++;
    if (pagesFetched % 10 === 0 || pagesFetched === pageQueue.length) {
      console.log(`  Progress: ${pagesFetched}/${pageQueue.length} pages, ${allPreprints.length} preprints`);
    }
  }

  // Process with concurrency limit
  while (pageQueue.length > 0 || activeRequests > 0) {
    while (activeRequests < CONCURRENCY && pageQueue.length > 0) {
      const page = pageQueue.shift();
      activeRequests++;
      processPage(page).catch(e => {
        console.error(`  Error on page ${page}: ${e.message}`);
      }).finally(() => {
        activeRequests--;
      });
    }
    if (pageQueue.length > 0 || activeRequests > 0) {
      await sleep(100);
    }
  }

  // Wait for any remaining requests
  await sleep(2000);

  // Save cache
  writeFileSync(CACHE_FILE, JSON.stringify(allPreprints, null, 2));
  console.log(`\nSaved ${allPreprints.length} preprints to ${CACHE_FILE}`);
  return allPreprints;
}

// Extract key words from a short title
function extractKeywords(shortTitle) {
  const stop = new Set([
    'a','an','the','and','or','of','for','in','on','to','with','by','from',
    'their','its','is','are','was','were','be','been','that','this','these',
    'those','not','no','but','at','if','study','analysis','review','research',
    'examination','based','using','between','among','into','clinical',
    'framework','model','effects','effect','role','association','predictors',
    'outcomes','findings','evidence','overview','perspective','considerations',
    'evaluating','evaluation','assessment','versus','therapeutic','approaches',
    'approach','systematic','additional','implications'
  ]);
  const words = shortTitle.match(/[a-zA-Z]+/g) || [];
  return words.filter(w => !stop.has(w.toLowerCase()) && w.length >= 3);
}

// Compute match score
function matchScore(shortTitle, apiTitle) {
  const sLower = shortTitle.toLowerCase();
  const aLower = apiTitle.toLowerCase();
  const keywords = extractKeywords(shortTitle);

  if (keywords.length === 0) return 0;

  // Key word hits
  const keyHits = keywords.filter(w => aLower.includes(w.toLowerCase())).length;
  const keyRatio = keyHits / keywords.length;

  // All word overlap
  const sWords = new Set(sLower.split(/\s+/));
  const aWords = new Set(aLower.split(/\s+/));
  const wordOverlap = [...sWords].filter(w => aWords.has(w) && w.length >= 3).length / sWords.size;

  // Rare term matching (words >= 7 chars are very specific)
  const rareTerms = keywords.filter(w => w.length >= 7);
  let rareBonus = 0;
  if (rareTerms.length > 0) {
    const rareHits = rareTerms.filter(w => aLower.includes(w.toLowerCase())).length;
    rareBonus = rareHits / rareTerms.length;
  }

  // Substring matching for specific phrases (e.g., "PEP therapy", "CBM-I")
  const phrases = shortTitle.match(/"[^"]+"|\([^)]+\)|[\w]+-[\w]+/g) || [];
  let phraseHits = 0;
  for (const phrase of phrases) {
    if (aLower.includes(phrase.toLowerCase().replace(/['"]/g, ''))) phraseHits++;
  }
  const phraseRatio = phrases.length > 0 ? phraseHits / phrases.length : 0;

  // Combined score
  let score;
  if (rareBonus >= 0.5 && rareTerms.length >= 2) {
    score = Math.max(keyRatio * 0.95, wordOverlap * 0.8, 0.65);
    score += rareBonus * 0.15;
  } else if (keyRatio >= 0.5) {
    score = Math.max(keyRatio * 0.9, wordOverlap * 0.7, phraseRatio * 0.8);
  } else if (keyRatio >= 0.3 && (wordOverlap > 0.3 || phraseRatio > 0)) {
    score = Math.max(keyRatio * 0.8, wordOverlap * 0.5, phraseRatio * 0.7);
  } else {
    score = Math.max(keyRatio * 0.7, wordOverlap * 0.4);
  }

  return Math.min(score, 1.0);
}

async function main() {
  const papers = JSON.parse(readFileSync(PAPERS_JSON, 'utf-8'));
  const missing = papers.filter(p => !p.osf_id);
  console.log(`Found ${missing.length} papers missing osf_id\n`);

  // Download all titles
  const allPreprints = await downloadAllTitles();
  console.log(`\nMatching against ${allPreprints.length} PsyArXiv preprints...\n`);

  const found = {};
  const notFound = [];

  for (const paper of missing) {
    const { number, title } = paper;
    let bestMatch = null;
    let bestScore = 0;

    for (const preprint of allPreprints) {
      const score = matchScore(title, preprint.title);
      if (score > bestScore) {
        bestScore = score;
        bestMatch = preprint;
      }
    }

    if (bestMatch && bestScore >= 0.35) {
      // Validate: at least one keyword must be in the API title
      const keywords = extractKeywords(title);
      const validHits = keywords.filter(w => bestMatch.title.toLowerCase().includes(w.toLowerCase())).length;
      if (validHits === 0 && keywords.length >= 2) {
        console.log(`#${number} REJECTED (no keywords match, score=${bestScore.toFixed(2)}): ${title.substring(0, 55)}`);
        console.log(`  Best: ${bestMatch.title.substring(0, 70)}`);
        notFound.push(number);
        continue;
      }

      const osfId = bestMatch.id.replace(/_v\d+$/i, '');
      console.log(`#${number} MATCH (${bestScore.toFixed(2)}): ${title.substring(0, 55)}`);
      console.log(`  -> ${bestMatch.title.substring(0, 70)}`);
      console.log(`  osf_id=${osfId}`);
      found[number] = {
        osf_id: osfId,
        api_title: bestMatch.title,
        link: `https://osf.io/preprints/psyarxiv/${osfId}`,
        score: +bestScore.toFixed(3),
        original_title: title,
      };
    } else {
      console.log(`#${number} NO MATCH (best=${bestScore.toFixed(2)}): ${title.substring(0, 55)}`);
      if (bestMatch) console.log(`  Best: ${bestMatch.title.substring(0, 70)}`);
      notFound.push(number);
    }
  }

  console.log(`\n${'='.repeat(60)}`);
  console.log(`RESULTS: Found ${Object.keys(found).length}/${missing.length} matches`);
  console.log(`${'='.repeat(60)}`);

  if (Object.keys(found).length > 0) {
    console.log(`\n--- FOUND ---`);
    for (const [num, info] of Object.entries(found).sort((a, b) => +a[0] - +b[0])) {
      console.log(`  #${num}: ${info.original_title.substring(0, 60)}`);
      console.log(`    -> ${info.api_title.substring(0, 70)}`);
      console.log(`    osf_id=${info.osf_id} score=${info.score}`);
    }
  }

  if (notFound.length > 0) {
    console.log(`\n--- NOT FOUND (${notFound.length}) ---`);
    for (const num of notFound) {
      const p = missing.find(m => m.number === num);
      console.log(`  #${num}: ${p.title}`);
    }
  }

  const output = {
    found,
    not_found_numbers: notFound,
    total_missing: missing.length,
    total_found: Object.keys(found).length,
  };
  writeFileSync(OUTPUT_FILE, JSON.stringify(output, null, 2));
  console.log(`\nResults saved to ${OUTPUT_FILE}`);
}

main().catch(err => { console.error(err); process.exit(1); });

File: scripts/find-missing-ids-bulk.py
#!/usr/bin/env python3
"""
Download ALL PsyArXiv preprint IDs and titles from the OSF API,
then match them against the 59 papers missing osf_ids.
This is more reliable than search-based approaches.
"""
import json
import urllib.request
import urllib.parse
import time
import sys
import os
from difflib import SequenceMatcher

PAPERS_JSON = "/home/z/my-project/psyarxiv-hub/data/papers.json"
API_BASE = "https://api.osf.io/v2/preprints/"
PAGE_SIZE = 100
CACHE_FILE = "/home/z/my-project/scripts/psyarxiv-all-titles.json"
# How many pages to fetch (10000+ preprints / 100 per page = ~100+ pages)
MAX_PAGES = 150

def api_get(url, retries=3):
    for attempt in range(retries):
        try:
            req = urllib.request.Request(url, headers={
                "Accept": "application/json",
                "User-Agent": "Mozilla/5.0 (compatible; PsyArXiv Matcher)"
            })
            with urllib.request.urlopen(req, timeout=30) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except Exception as e:
            if attempt < retries - 1:
                time.sleep(2 ** attempt)
            else:
                print(f"  API error for {url}: {e}", file=sys.stderr)
                return None

def download_all_titles():
    """Download all PsyArXiv preprint titles."""
    if os.path.exists(CACHE_FILE):
        print(f"Loading cached titles from {CACHE_FILE}")
        with open(CACHE_FILE) as f:
            return json.load(f)
    
    print("Downloading all PsyArXiv preprint titles...")
    all_preprints = []
    
    for page in range(1, MAX_PAGES + 1):
        params = urllib.parse.urlencode({
            "filter[provider]": "psyarxiv",
            "page[size]": PAGE_SIZE,
            "page[number]": page,
            "sort": "-date_created",
        })
        url = f"{API_BASE}?{params}"
        data = api_get(url)
        if not data:
            print(f"  Failed at page {page}, stopping")
            break
        
        items = data.get("data", [])
        if not items:
            print(f"  No items at page {page}, stopping")
            break
        
        for item in items:
            attrs = item.get("attributes", {})
            all_preprints.append({
                "id": item["id"],
                "title": attrs.get("title", ""),
                "date_created": attrs.get("date_created", "")[:10],
            })
        
        total = data.get("meta", {}).get("total", "?")
        print(f"  Page {page}: fetched {len(items)} ({len(all_preprints)} total / {total})")
        
        # Check if there's a next page
        next_url = data.get("links", {}).get("next")
        if not next_url:
            print(f"  No next page, done at page {page}")
            break
        
        time.sleep(0.3)  # Rate limiting
    
    # Save cache
    with open(CACHE_FILE, "w") as f:
        json.dump(all_preprints, f)
    print(f"Saved {len(all_preprints)} preprints to {CACHE_FILE}")
    return all_preprints

def extract_key_phrases(short_title):
    """Extract key phrases/words from a short title for matching."""
    stop = {
        'a', 'an', 'the', 'and', 'or', 'of', 'for', 'in', 'on', 'to', 'with',
        'by', 'from', 'their', 'its', 'is', 'are', 'was', 'were', 'be', 'been',
        'that', 'this', 'these', 'those', 'not', 'no', 'but', 'at', 'if',
        'study', 'analysis', 'review', 'research', 'examination', 'based',
        'using', 'between', 'among', 'into', 'clinical', 'framework', 'model',
        'effects', 'effect', 'role', 'association', 'predictors', 'outcomes',
        'findings', 'evidence', 'overview', 'perspective', 'considerations',
        'evaluating', 'evaluation', 'assessment', 'versus', 'additional',
        'therapeutic', 'approaches', 'approach', 'systematic',
    }
    words = [w.lower() for w in __import__('re').findall(r'[a-zA-Z]+', short_title)]
    key_words = [w for w in words if w not in stop and len(w) >= 3]
    return key_words, words

def compute_match_score(short_title, api_title, paper_authors="Unknown"):
    """Compute a match score between a short title and an API title."""
    short_lower = short_title.lower()
    api_lower = api_title.lower()
    
    # 1. Fuzzy sequence similarity
    seq_sim = SequenceMatcher(None, short_lower, api_lower).ratio()
    
    # 2. Key word overlap
    key_words, all_words = extract_key_phrases(short_title)
    if not key_words:
        return 0, "no key words"
    
    key_hits = sum(1 for w in key_words if w in api_lower)
    key_ratio = key_hits / len(key_words)
    
    # 3. All word overlap (less strict)
    all_hits = sum(1 for w in all_words if w in api_lower)
    all_ratio = all_hits / len(all_words) if all_words else 0
    
    # 4. Check for rare/specific terms (longer words = more specific)
    rare_terms = [w for w in key_words if len(w) >= 7]
    rare_hits = sum(1 for w in rare_terms if w in api_lower)
    rare_ratio = rare_hits / len(rare_terms) if rare_terms else 0
    
    # Combined score
    # Rare term match is very strong signal
    if rare_ratio >= 0.5 and rare_terms:
        score = max(seq_sim, key_ratio * 0.95, 0.7)
    elif key_ratio >= 0.6:
        score = max(seq_sim, key_ratio * 0.9, 0.6)
    elif key_ratio >= 0.4 and seq_sim > 0.2:
        score = max(seq_sim * 0.8, key_ratio * 0.7)
    else:
        score = max(seq_sim, key_ratio * 0.6, all_ratio * 0.5)
    
    return score, f"seq={seq_sim:.2f} key={key_ratio:.2f} all={all_ratio:.2f} rare={rare_ratio:.2f}"

def find_matches(all_preprints, missing_papers):
    """Find the best match for each missing paper."""
    results = {}
    
    for paper in missing_papers:
        number = paper["number"]
        title = paper["title"]
        authors = paper.get("authors", "Unknown")
        
        best_match = None
        best_score = 0
        best_details = ""
        
        for preprint in all_preprints:
            score, details = compute_match_score(title, preprint["title"])
            if score > best_score:
                best_score = score
                best_match = preprint
                best_details = details
        
        if best_match and best_score >= 0.40:
            osf_id = best_match["id"].rsplit("_", 1)[0]
            print(f"  #{number} (score={best_score:.2f}): {title[:50]}")
            print(f"    -> {best_match['title'][:70]}")
            print(f"    osf_id={osf_id} | {best_details}")
            results[number] = {
                "osf_id": osf_id,
                "api_title": best_match["title"],
                "link": f"https://osf.io/preprints/psyarxiv/{osf_id}",
                "score": round(best_score, 3),
                "details": best_details,
            }
        else:
            print(f"  #{number} NO MATCH (best={best_score:.2f}): {title[:50]}")
            if best_match:
                print(f"    Best: {best_match['title'][:70]} | {best_details}")
            results[number] = None
    
    return results

def main():
    # Load papers
    with open(PAPERS_JSON) as f:
        papers = json.load(f)
    
    missing = [p for p in papers if not p.get("osf_id")]
    print(f"Found {len(missing)} papers missing osf_id\n")
    
    # Download all titles
    all_preprints = download_all_titles()
    print(f"\nMatching against {len(all_preprints)} PsyArXiv preprints...\n")
    
    # Find matches
    matches = find_matches(all_preprints, missing)
    
    # Summary
    found = {k: v for k, v in matches.items() if v is not None}
    not_found = [k for k, v in matches.items() if v is None]
    
    print(f"\n{'='*60}")
    print(f"RESULTS: Found {len(found)}/{len(missing)} matches")
    print(f"{'='*60}")
    
    # Save results
    output = {
        "found": found,
        "not_found_numbers": not_found,
        "total_missing": len(missing),
        "total_found": len(found),
    }
    out_path = "/home/z/my-project/scripts/missing-osf-results.json"
    with open(out_path, "w") as f:
        json.dump(output, f, indent=2)
    print(f"\nResults saved to {out_path}")
    
    # Print found matches for review
    if found:
        print(f"\n--- FOUND MATCHES ---")
        for num, info in sorted(found.items()):
            p = next(p for p in papers if p["number"] == num)
            print(f"  #{num}: {p['title'][:60]}")
            print(f"    -> {info['api_title'][:70]}")
            print(f"    osf_id={info['osf_id']} score={info['score']}")
            print()

if __name__ == "__main__":
    main()